application/src/main/java/org/opentripplanner/updater/trip/TripPatternCache.java

```diff
@@ -0,0 +1,107 @@
+package org.opentripplanner.updater.trip;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import org.opentripplanner.transit.model.network.StopPattern;
+import org.opentripplanner.transit.model.network.TripPattern;
+import org.opentripplanner.transit.model.timetable.Trip;
+
+/**
+ * Thread-safe mechanism for tracking any TripPatterns added to the graph via realtime messages.
+ * This tracks only patterns added by realtime messages, not ones that already existed from the
+ * scheduled data. This is a "cache" in the sense that it will keep returning the same TripPattern
+ * when presented with the same StopPattern, so if realtime messages add many trips passing through
+ * the same sequence of stops, they will all end up on this same TripPattern.
+ *
+ * <h3>Thread-safety</h3>
+ *
+ * The thread safety is guaranteed by only a single thread being allowed to access this cache through
+ * {@link org.opentripplanner.updater.GraphWriterRunnable}. This, however, also means that a thread
+ * could modify the underlying snapshot buffer and then immediately afterwards access this cache
+ * which has access to the very same snapshot.
+ * <p>
+ * <a href="https://github.com/opentripplanner/OpenTripPlanner/pull/6629#pullrequestreview-2854296598">Further reading</a>
```

Comment:
Thanks, this is basically what I wanted (make sure the observations I made are readily reachable if/when this is refactored in the future). The sentence about problematic access patterns by multiple threads is hard for me to understand, it would probably have to be much longer to really describe the problem. I suggest we leave it out and only link to my review comment:

```suggestion
 * {@link org.opentripplanner.updater.GraphWriterRunnable}. 
 * <a href="https://github.com/opentripplanner/OpenTripPlanner/pull/6629#pullrequestreview-2854296598">Further reading</a>
```